content/blog/copilot-lessons/index.md

```diff
@@ -0,0 +1,131 @@
+---
+title: "Engineering Lessons from Building Pulumi Copilot"
+date: 2024-12-12T13:56:13-05:00
+draft: false
+meta_desc: Learn key engineering lessons from building Pulumi Copilot, including how to minimize LLM workload, validate outputs, and deal with hallucination.
+meta_image: meta.png
+authors:
+    - artur-laksberg
+    - adam-gordon-bell
+tags:
+    - copilot
+    - ai
+    - iac
+social:
+    twitter: 🚀 Building Pulumi Copilot taught us key AI engineering insights – Minimize LLM workload, validate outputs rigorously & use "skills" for modular tasks. Plus, sometimes when AI hallucinates features, it reveals exactly what users want! 
+    linkedin: 🚀 Building Pulumi Copilot taught us key AI engineering insights – Minimize LLM workload, validate outputs rigorously & use "skills" for modular tasks. Plus, sometimes when AI hallucinates features (like a "--force" flag), it reveals exactly what users want! 
+---
+**Artur Laksberg is on the Copilot team here at Pulumi. CoPilot is an AI assistant for cloud infrastructure. Today he shares key lessons from bringing Copilot to production—including an unexpected discovery about AI hallucinations.**
+
+While reviewing user feedback for Pulumi Copilot, our AI assistant for cloud infrastructure, one message caught my eye: "Your tool doesn't know anything!". Having just made some changes, I braced for the worst. But the evals were still looking strong, so what was going on?
+
+<!--more-->
+
+The user was trying to force-delete a stack that still had resources. Copilot confidently suggested using a `--force` flag, which is a perfectly logical solution, except this flag doesn't exist in Pulumi. This was pure hallucination!
+
+Welcome to a day in the life of building with LLMs. But this particular error, this is a specific type of hallucination that I now have a solution for. And now that [we've launched our REST API](/blog/pulumi-copilot-rest/) and Copilot has been running in production for a bit, **I want to share some lessons we've learned building with LLMs**.
```

Comment:
> I now have a solution for

We have not made a decision add that flag yet. I would just say that Copilot hallucinated something that isn't unreasonable, especially given that `pulumi stack rm` command does have that flag. (https://www.pulumi.com/docs/iac/cli/commands/pulumi_stack_rm/)

Reply:
I meant it more conceptually, hence not "this specific hallucination" but "this specific type of hallucination" and solution is to consider it as potential product feedback.